src/store/indexeddb.js

```diff
@@ -0,0 +1,403 @@
+        this._userModifiedMap[u.userId] = u.getLastModifiedTime();
+    });
+
+    // TODO: work out changed account data events. They don't have timestamps or IDs.
+    // so we'll need to hook into storeAccountDataEvents instead to catch them when
+    // they update from /sync
+    const changedAccountData = Object.keys(this.accountData).map((etype) => {
+        return this.accountData[etype];
+    });
+
+    const syncData = this._syncAccumulator.getJSON();
+
+    return q.all([
+        this.backend.persistUsers(changedUsers),
+        this.backend.persistAccountData(changedAccountData),
+        this.backend.persistSyncData(syncData.nextBatch, syncData.roomsData),
+    ]);
+};
+
+function createDatabase(db) {
+    // Make user store, clobber based on user ID. (userId property of User objects)
+    db.createObjectStore("users", { keyPath: ["userId"] });
+
+    // Make account data store, clobber based on event type.
+    // (event.type property of MatrixEvent objects)
+    db.createObjectStore("accountData", { keyPath: ["type"] });
+
+    // Make /sync store (sync tokens, room data, etc), always clobber (const key).
+    db.createObjectStore("sync", { keyPath: ["clobber"] });
+}
+
+/**
+ * Helper method to collect results from a Cursor and promiseify it.
+ * @param {ObjectStore|Index} store The store to perform openCursor on.
+ * @param {IDBKeyRange=} keyRange Optional key range to apply on the cursor.
+ * @param {Function} resultMapper A function which is repeatedly called with a
+ * Cursor.
+ * Return the data you want to keep.
+ * @return {Promise<T[]>} Resolves to an array of whatever you returned from
+ * resultMapper.
+ */
+function selectQuery(store, keyRange, resultMapper) {
+    const query = store.openCursor(keyRange);
+    return q.Promise((resolve, reject) => { /*eslint new-cap: 0*/
+        const results = [];
+        query.onerror = (event) => {
+            reject(new Error("Query failed: " + event.target.errorCode));
+        };
+        // collect results
+        query.onsuccess = (event) => {
+            const cursor = event.target.result;
+            if (!cursor) {
+                resolve(results);
+                return; // end of results
+            }
+            results.push(resultMapper(cursor));
+            cursor.continue();
+        };
+    });
+}
+
+function promiseifyTxn(txn) {
+    return new q.Promise((resolve, reject) => {
+        txn.oncomplete = function(event) {
+            resolve(event);
+        };
+        txn.onerror = function(event) {
+            reject(event);
+        };
+    });
+}
+
+function promiseifyRequest(req) {
+    return new q.Promise((resolve, reject) => {
+        req.onsuccess = function(event) {
+            resolve(event);
+        };
+        req.onerror = function(event) {
+            reject(event);
+        };
+    });
+}
+
+module.exports.IndexedDBStore = IndexedDBStore;
+module.exports.IndexedDBStoreBackend = IndexedDBStoreBackend;
```

Comment:
We do this because of JSDoc: https://github.com/matrix-org/matrix-js-sdk/pull/363#discussion_r101494996